x-pack/plugins/security/public/authentication/login/login_page.tsx

```diff
@@ -144,68 +175,120 @@ export class LoginPage extends Component<Props, State> {
+            />
+          }
+          message={
+            <FormattedMessage
+              id="xpack.security.loginPage.xpackUnavailableMessage"
+              defaultMessage="To use the full set of free features in this distribution of Kibana, please update Elasticsearch to the default distribution."
+            />
+          }
+        />
+      );
+    }
+
+    if (layout !== 'form') {
+      return (
+        <DisabledLoginForm
+          title={
+            <FormattedMessage
+              id="xpack.security.loginPage.unknownLayoutTitle"
+              defaultMessage="Unsupported login form layout."
+            />
+          }
+          message={
+            <FormattedMessage
+              id="xpack.security.loginPage.unknownLayoutMessage"
+              defaultMessage="Refer to the Kibana logs for more details and refresh to try again."
+            />
+          }
+        />
+      );
+    }
+
+    const loginSelector =
+      showLoginSelector &&
+      selector.providers.map((provider, index) => (
+        <EuiButton
+          key={index}
+          className="loginWelcome__selectorButton"
+          iconType="user"
+          fullWidth={true}
+          onClick={() => this.login(provider.type, provider.name)}
+        >
+          {provider.options.description ?? `${provider.type}/${provider.name}`}
+        </EuiButton>
+      ));
+
+    const loginSelectorAndLoginFormSeparator = showLoginSelector && showLoginForm && (
+      <>
+        <EuiText textAlign="center" color="subdued">
+          ―――&nbsp;&nbsp;
+          <FormattedMessage id="xpack.security.loginPage.loginSelectorOR" defaultMessage="OR" />
+          &nbsp;&nbsp;―――
+        </EuiText>
+        <EuiSpacer size="m" />
+      </>
+    );
+
+    const loginForm = showLoginForm && (
+      <BasicLoginForm
+        http={this.props.http}
+        infoMessage={infoMessageMap.get(parse(window.location.href, true).query.msg?.toString())}
+        loginAssistanceMessage={this.props.loginAssistanceMessage}
+      />
+    );
+
+    return (
+      <>
+        {loginSelector}
+        {loginSelectorAndLoginFormSeparator}
+        {loginForm}
+      </>
+    );
+  };
+
+  private login = async (providerType: string, providerName: string) => {
+    try {
+      const { location } = await this.props.http.post<{ location: string }>(
+        `${this.props.http.basePath.serverBasePath}/internal/security/login_with`,
+        { body: JSON.stringify({ providerType, providerName, currentURL: window.location.href }) }
+      );
+
+      window.location.href = location;
+    } catch (err) {
+      this.props.notifications.toasts.addDanger(
+        i18n.translate('xpack.security.loginPage.loginSelectorErrorMessage', {
+          defaultMessage: 'Could not perform login: {message}. Contact your system administrator.',
+          values: { message: (err as IHttpFetchError).message },
+        })
+      );
```

Comment:
__note:__ I'm thinking about simplifying this just to this:
```suggestion
      this.props.notifications.toasts.addError(err, {
        title: i18n.translate('xpack.security.loginPage.loginSelectorErrorMessage', {
          defaultMessage: 'Could not perform login.',
        }),
      });
```